astropy/modeling/spline.py

```diff
@@ -0,0 +1,438 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+"""Spline models and fitters."""
+# pylint: disable=line-too-long, too-many-lines, too-many-arguments, invalid-name
+
+import warnings
+
+import abc
+
+from astropy.utils.exceptions import (AstropyUserWarning,)
+from .core import (Fittable1DModel, Fittable2DModel, ModelDefinitionError)
+
+from .fitting import _FitterMeta
+
+
+class _Spline(abc.ABC):
+    """
+    Abstract Class for Splines
+
+    Parameters
+    ----------
+    optional_inputs: dict
+        dictionary of the form
+            optional input name: default value of input
+    """
+
+    optional_inputs = {}
+
+    @property
+    def spline(self):
+        return self._spline
+
+    @spline.setter
+    def spline(self, value):
+        if self._spline is not None:
+            warnings.warn("Spline already defined for this model, you are overriding it.",
+                          AstropyUserWarning)
+        self._spline = value
+
+    def reset_spline(self):
+        self._spline = None
+
+    @property
+    def bbox(self):
+        try:
+            return self.bounding_box.bbox
+        except NotImplementedError:
+            return [None]*(2*self.n_inputs)
+
+    @staticmethod
+    def _optional_arg(arg):
+        return f'_{arg}'
+
+    def _create_optional_inputs(self):
+        for arg in self.optional_inputs:
+            attribute = self._optional_arg(arg)
+            if hasattr(self, attribute):
+                raise ValueError(f'Optional argument {arg} already exists in this class!')
+            else:
+                setattr(self, attribute, None)
+
+    def _intercept_optional_inputs(self, **kwargs):
+        new_kwargs = kwargs
+        for arg in self.optional_inputs:
+            if (arg in kwargs):
+                attribute = self._optional_arg(arg)
+                if getattr(self, attribute) is None:
+                    setattr(self, attribute, kwargs[arg])
+                    del new_kwargs[arg]
+                else:
+                    raise RuntimeError(f'{arg} has already been set, something has gone wrong!')
+
+        return new_kwargs
+
+    def _get_optional_inputs(self, **kwargs):
+        optional_inputs = kwargs
+        for arg in self.optional_inputs:
+            attribute = self._optional_arg(arg)
+
+            if arg in kwargs:
+                # Options passed in
+                optional_inputs[arg] = kwargs[arg]
+            elif getattr(self, attribute) is not None:
+                # No options passed in and Options set
+                optional_inputs[arg] = getattr(self, attribute)
+                setattr(self, attribute, None)
+            else:
+                # No options passed in and No options set
+                optional_inputs[arg] = self.optional_inputs[arg]
+
+        return optional_inputs
+
+
+class Spline1D(Fittable1DModel, _Spline):
```

Comment:
There will need to be a way of creating a spline from from its internal parameters if we want to serialize it and deserialize it. I think that is missing now since it only appears to be the result of a fit. An it would be good to understand how the state is stored and can be accessed (i.e., methods or properties to obtain this).

Reply:
Currently the 1D spline is constructed either as an LSQUnivariateSpline or UnivariateSpline object, depending on whether knots are supplied. I would suggest that the spline is always a BSpline object (so only this class needs to be serialzed), which is most easily done by using the scipy.interpolate.splrep function to perform the spline-fitting, which returns the tck tuple that can be fed directly into the BSpline constructor.

Reply:
This could be more workable.

Reply:
Note that we may want to explicitly store the residual (square sum of the residuals) that `splrep` returns in addition to the spline itself.

Reply:
Refactored code to use the `splrep` function to generate a tck tuple, which is what is wrapped by the model.